### DISCUSSION TITLE
[Question] how to make Executable file with [Navigation and routing]??

### DISCUSSION THREAD

**Discussion by jjjunehkim** [flet-dev/flet#3101] [Q&A]
### Question

The desired execution process is as follows.

After being created  a Executable file(.exe file),
Run a Executable file(.exe file) on "A"computer(PC)
And connect "http://192.168.35.35:8550/data_table on "B" computer(PC) 

So, I am trying to make a Executable file(.exe file) through the process below. 
Could you help me?

1. Use this command -> flet pack myapp.py  (myapp.py : scripts name)
2. A executable file(myapp.exe) was created in the dist folder.
3. Run a myapp.exe
4. Error window opened
![flet_error](https://github.com/flet-dev/flet/assets/126040143/8b28d906-32b0-42be-b340-d7cb350a168e)

----------------------------------------------------------------------------------
 And I tried other method "flet build".  after change file name (myapp.py -> main.py) for using the command("flet build windows")
->>> command : flet build windows 
<result >
Creating Flutter bootstrap project...OK
Customizing app icons and splash images...OK
Generating app icons...OK
Packaging Python app...OK
Building Windows app...OK
Copying build to build\windows directory...OK
Success!

And......
I run the Executable file(.exe file) on "A"computer. it work. 
and I opened chorome browser.
and I try to connect  <"http://192.168.35.35:8550/data_table" in the chrome browser address bar>on "B" computer(PC)
but it was not opened on "B"computer.
 --------------------------------------------------------------------------------

One more trial!!!!
 And I tried other method "flet build".  after change file name (myapp.py -> main.py) for using the command("flet build web")
->>> command : flet build web
 
It also didn't work as expected.
 --------------------------------------------------------------------------------

### Code sample

```python
import flet as ft

def main(page: ft.Page):

    data_table_sample=ft.DataTable(
                        width=400,
                        border=ft.border.all(2, "red"),
                        border_radius=10,
                        vertical_lines=ft.border.BorderSide(3, "blue"),
                        horizontal_lines=ft.border.BorderSide(1, "green"),
                        sort_column_index=0,
                        sort_ascending=True,
                        heading_row_color=ft.colors.BLACK12,
                        heading_row_height=100,
                        data_row_color={"hovered": "0x30FF0000"},
                        show_checkbox_column=True,
                        divider_thickness=0,
                        column_spacing=200,

                        columns=[
                            ft.DataColumn(
                                ft.Text("Column 1"),),
                            ft.DataColumn(
                                ft.Text("Column 2"),
                                tooltip="This is a second column",
                                numeric=True,
                        ),
                        ],

                        rows=[
                            ft.DataRow([ft.DataCell(ft.Text("A")), ft.DataCell(ft.Text("1"))]),
                        ],
    )


    # ------------------------------------------------------------------------------------------------
    def add_data(e): 
        page.pubsub.send_all(e)
        page.update()

    def add_data_12(): 

        row = ft.DataRow(
            cells=[
                ft.DataCell(ft.Text("B")),
                ft.DataCell(ft.Text("2"))
                    ],
        )

        return row
    def on_message(row):
        row = add_data_12()
        data_table_sample.rows.append(row)
        page.update()


    page.pubsub.subscribe(on_message)

    def add_data_2(e):
        data_table_sample.rows.append(
            ft.DataRow(
                cells=[
                    ft.DataCell(ft.Text("C")),
                    ft.DataCell(ft.Text("3"))
                ],
            )
        )
        print("appended data_2 !!")
        page.update()
    # ------------------------------------------------

    def route_change(route):
        global message
        page.views.clear()
        page.views.append(
            ft.View(
                "/",
                [
                    ft.Text("Main"),
                    ft.ElevatedButton(text="add_data", on_click = add_data, data = "add_data"),
                    data_table_sample
                ],
            )
        )
        if page.route == "/data_table":
            page.views.append(
                ft.View(
                    "/data_table",
                    [
                        ft.Text("Serve"),
                        ft.ElevatedButton(text="add_data", on_click = add_data, data = "add_data"),
                        data_table_sample
                    ],
                )
            )
        page.update()

    def view_pop(view):
        page.views.pop()
        top_view = page.views[-1]
        page.go(top_view.route)


    page.on_route_change = route_change
    page.on_view_pop = view_pop
    page.go(page.route)


ft.app(target=main, host="192.168.35.35", port= 8550, view=ft.AppView.WEB_BROWSER)


# host="192.168.35.35" --> it is my ip-address
```

### Error message

```bash
Traceback (most recent call last):
  File "logging\config.py", line 560, in configure
  File "logging\config.py", line 672, in configure_formatter
  File "logging\config.py", line 490, in configure_custom
  File "uvicorn\logging.py", line 42, in __init__
AttributeError: 'NoneType' object has no attribute 'isatty'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "myapp.py", line 144, in <module>
  File "flet_runtime\app.py", line 70, in app
  File "asyncio\runners.py", line 194, in run
  File "asyncio\runners.py", line 118, in run
  File "asyncio\base_events.py", line 685, in run_until_complete
  File "flet_runtime\app.py", line 156, in app_async
  File "flet_runtime\app.py", line 292, in __run_web_server
  File "flet\fastapi\serve_fastapi_web_app.py", line 85, in serve_fastapi_web_app
  File "uvicorn\config.py", line 272, in __init__
  File "uvicorn\config.py", line 364, in configure_logging
  File "logging\config.py", line 914, in dictConfig
  File "logging\config.py", line 563, in configure
ValueError: Unable to configure formatter 'default'
```

### ------------------------------------------------------

- [X] I have searched for answers to my question both in the issues and in previous discussions.

**Comment by ingdesarr1** [comment]
I use auto-py-to-exe and works fine, maybe you can try it.

**Reply by jjjunehkim** [reply]
Thank you for your answer!

I tried it.
But it still doesn't work.

I installed and execute a auto-py-to-exe.
when I run auto-py-to-exe on teminal,
and it occurs error(?) message
noticed  -> "warning: Installing shim for bottle import redirects (using Python=3.12.2 and bottle=0.12.25)"
It was caused error?
 
anyway a Executable file(.exe file) was  created . 
I run a .exe file.
and I got a error messase windows.

<error message>
Traceback (most recent call last):
  File "logging\config.py", line 560, in configure
  File "logging\config.py", line 672, in configure_formatter
  File "logging\config.py", line 490, in configure_custom
  File "uvicorn\logging.py", line 42, in __init__
AttributeError: 'NoneType' object has no attribute 'isatty'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "main.py", line 144, in <module>
  File "flet_runtime\app.py", line 70, in app
  File "asyncio\runners.py", line 194, in run
  File "asyncio\runners.py", line 118, in run
  File "asyncio\base_events.py", line 685, in run_until_complete
  File "flet_runtime\app.py", line 156, in app_async
  File "flet_runtime\app.py", line 292, in __run_web_server
  File "flet\fastapi\serve_fastapi_web_app.py", line 85, in serve_fastapi_web_app
  File "uvicorn\config.py", line 272, in __init__
  File "uvicorn\config.py", line 364, in configure_logging
  File "logging\config.py", line 914, in dictConfig
  File "logging\config.py", line 563, in configure
ValueError: Unable to configure formatter 'default'


![20240430_error](https://github.com/flet-dev/flet/assets/126040143/131f1c3c-569c-4d36-981d-9cfa3bcd3e60)

**Reply by ingdesarr1** [reply]
in your project, you use any other file or dependence? maybe you are not attach to project. what other files you use?

**Reply by jjjunehkim** [reply]
I executed the code that I posted when I asked the first question.
The one at the top is the sample code I ran.

**Reply by ndonkoHenri** [reply]
Which windows version are you using exactly ?

**Reply by jjjunehkim** [reply]
Version :   Window 11 pro 22H2

Is it related to the Windows version?

**Reply by jjjunehkim** [reply]
I will explain it in detail again 

------------------------------------------------------------------
I hope to be executed like either "_first case_" or "_second case_" when I run a executable file(.exe file)
------------------------------------------------------------------
_1. first case_
(<flet run> command in a command terminal )
after then,
-> When you click the <add_data> button for "A PC" and the <add_data> button for "B PC", data table rows for both "A PC" and "B PC" are added.

Image  executed only as an app when executing the <flet run> command in the command terminal of "A" PC

![flet_run_A_PC](https://github.com/flet-dev/flet/assets/126040143/b96fec0e-4716-4b58-8699-ac3bcc8d06a4)

Image opened in the chrome browser of "B" PC on the same network after running the <flet run> command in the command terminal of "A" PC

![flet_run_B_PC](https://github.com/flet-dev/flet/assets/126040143/1c78d005-7d79-4a5a-a3cc-e7500560078f)


_2..Second case_ 
(<(python main.py> command in a command terminal )
after then,
-> When you click the <add_data> button for "A PC" and the <add_data> button for "B PC", data table rows for both "A PC" and "B PC" are added.
Image opened in Chrome browser after executing the <python main.py> command in the command terminal of "A PC"
![A_PC](https://github.com/flet-dev/flet/assets/126040143/b4653767-a746-4967-9415-fc0c20cf54fd)

Image opened in the chrome browser of "B" PC on the same network after executing the <python main.py> command in the command terminal of "A" PC

![python_B_PC](https://github.com/flet-dev/flet/assets/126040143/51d190b2-2dfe-4b98-83df-237cb9ea010b)

=================================================================================

-----------------------------------------------------------------------------------------
Result :
 run a executable file(.exe file) created as the <flet build window> command
after then,
->   When you click the <add_data> button for "A PC" , data table rows for "A PC" are added.
     HTTP ERROR 404  in Chrome browser of "B" PC
----------------------------------------------------------------------------------------
Image when executing the executable file created with the <flet build window> command on "A" PC

![flet_run_A_PC](https://github.com/flet-dev/flet/assets/126040143/794b5a7a-384f-440f-a65a-ac61cfc4bd97)


The image opened in the Chrome browser on "B" PC, after executing the executable file created with the <flet build window> command on "A" PC. 

![B_PC](https://github.com/flet-dev/flet/assets/126040143/db2d69fd-5188-47f9-b82a-13ef99bcb5d6)
